core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -2790,13 +2790,13 @@ public static double log(BigDecimal d0, BigDecimal d1) {
 
   /** SQL {@code LOG2(number)} function applied to double values. */
   public static double log2(double d0) {
-    return (d0 == 0) ? Double.NaN : Math.log(d0) / Math.log(2);
+    return (Double.isInfinite(log(d0, 2))) ? Double.NaN : log(d0, 2);
```

Comment:
I don't think this will work for the case of +infinity.
You should add a test case for `log2(CAST "+Infinity' AS NULL)`. I expect that the result will be NaN, whereas you probably expect Infinity in that case. So you should also check that the infinity is negative here.

Reply:
that should work using < 0.0.

Reply:
> I don't think this will work for the case of +infinity. You should add a test case for `log2(CAST "+Infinity' AS NULL)`. I expect that the result will be NaN, whereas you probably expect Infinity in that case. So you should also check that the infinity is negative here.

Maybe I should check how mysql and spark work

Reply:
> that should work using < 0.0.

The result of log2(2.0/3) is also less than 0.0，

Reply:
`Double.isInfinite(d0) && d0 < 0`

Reply:
Thank you

Reply:
wait， Double.isInfinite(d0) && d0 < 0 ， I think you mean Double.isInfinite(log(d0, 2)) && d0.doubleValue() < 0.0

Reply:
just make sure it matches MySQL on all corner cases

Reply:
cast as double

Reply:
Mysql seems to be unable to support the parsing of +Infinity and -Infinity. From a mathematical point of view, I am not sure whether this makes sense, but mysql obviously does not support it. Maybe we can only deal with 0 specifically. I tested +0.0 and - 0.0 returns null

Reply:
you can try a very large value, such as 1e+52
https://dev.mysql.com/doc/refman/8.0/en/problems-with-float.html

Reply:
Thank you very much for answering my question so patiently